### PR TITLE
ASI-209: [MFTF]: User searches stock images by keywords

### PR DESCRIPTION
**Pull request by engcom-Golf**
### Description (*)
Add functional test.

### Fixed Issues (if relevant)
1. magento/adobe-stock-integration#209: [MFTF]: User searches stock images by keywords

### Manual testing scenarios (*)
1. Run functional test Magento/AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml

### CODE REVIEW

File: AdobeStockImageAdminUi/view/adminhtml/web/template/grid/toolbar.html [hunk 1]
```diff
@@ -4,7 +4,7 @@
  * See COPYING.txt for license details.
  */
 -->
-<div class="admin__data-grid-header" afterRender="$data.setToolbarNode">
+<div class="admin__data-grid-header masonry-main-toolbar" afterRender="$data.setToolbarNode">
```
Comment (hws47a) [inline AdobeStockImageAdminUi/view/adminhtml/web/template/grid/toolbar.html]: Classes are generally used for styling and are not recommended to use as a selectors in javascript or UI tests.  
I would suggest to use data-role attribute or create a specific attribute which can be used only for tests.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        </annotations>
+        <before>
+            <actionGroup ref="LoginActionGroup" stepKey="loginGetFromGeneralFile"/>
+        </before>
+        <after>
+            <amOnPage url="admin/admin/auth/logout/" stepKey="amOnLogoutPage"/>
+        </after>
+        <amOnPage url="{{CmsPagesPage.url}}" stepKey="amOnPagePagesGrid"/>
+        <waitForPageLoad stepKey="waitForPageLoad"/>
+        <click selector="{{CmsPagesPageActionsSection.addNewPageButton}}" stepKey="clickAddNewPage"/>
```
Comment (hws47a) [inline AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: The first few lines of the tests duplicate the same in another test.  
It's better to create an action group to avoid duplicates.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        <grabAttributeFrom selector="div[class='masonry-image-column'][data-repeat-index='0'] img" userInput="src" stepKey="getUrlFromFirstImageAfterSearchChestBand"/>
+        <fillField selector="{{AdobeStockSection.searchInput}}" userInput="mi band heart rate" stepKey="searchForMiBand"/>
+        <click selector="{{AdobeStockSection.searchButton}}" stepKey="applySearchWithMiBand"/>
+        <waitForPageLoad stepKey="waitForSearchWithMiBand"/>
+        <dontSeeElement selector="//div[@class='masonry-image-column' and contains(@data-repeat-index, '0')]//img[@src='{$getUrlFromFirstImageAfterSearchChestBand}']" stepKey="checkFirstImageNotFoundAfterSearchMiBand"/>
+        <dontSeeElement selector="//aside[contains(@class, 'modal-popup')]//div[@class='modal-content']//div[text()='Something went wrong.']" stepKey="errorModalIsAbsentAfterSearchChestBand"/>
+        <dontSeeElement selector="//div[contains(@id, 'adobe-stock-images-search-modal')]//div[@class='spinner']" stepKey="spinnerIsAbsentAfterSearchChestBand"/>
```
Comment (hws47a) [inline AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: Test in general looks two complex and contains too many steps. I would suggest create a few meaningful action groups and use them instead.

File: AdobeStockClient/Model/Client.php [hunk 4]
```diff
@@ -25,7 +25,6 @@
 use Magento\Framework\Api\SearchCriteriaInterface;
 use Magento\Framework\Api\Search\SearchResultInterface;
 use Magento\Framework\Api\Search\SearchResultFactory;
-use Magento\Framework\Api\SearchCriteriaInterface;
```
Comment (hws47a) [inline AdobeStockClient/Model/Client.php]: Doesn't seem related to this test

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 2 above)
Comment (hws47a) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: Ones merged, action group from this pull request can be used
https://github.com/magento/adobe-stock-integration/pull/239

**Comment by engcom-Golf** [comment]
Related to pull request #239

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 2 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: @hws47a Hello. In your PR (239) you created action group "AdminOpenAdobeStockSearchModalActionGroup" and in one of the steps you use selector TinyMCESection.InsertImageIcon, it doesn't work for me and I didn't found class .mce-i-image on page

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (rogyar) [inline AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml]: According to the best practices in MFTF we should use one action group per file.

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AdminOpenAdobeStockSearchModalViaGalleryButtonActionGroup">
```
Comment (rogyar) [inline AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml]: Just for note, we need to make sure that we are not duplicating the action groups with different names but same actions. Please check the action group introduced here https://github.com/magento/adobe-stock-integration/pull/239/files#diff-f9da93ddc7a90bd437d0174ab9c8ded2R11

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            <features value="AdobeStockImagePanel"/>
+            <stories value="Access Adobe Stock Image panel via the Admin"/>
+            <title value="User searches stock images by keywords"/>
+            <description value="User searches stock images by keywords"/>
+            <severity value="CRITICAL"/>
+            <group value="AdobeStockIntegration"/>
+        </annotations>
+        <before>
+            <actionGroup ref="LoginAsAdmin" stepKey="loginAsAdmin"/>
+            <actionGroup ref="AdminAdobeStockGoToConfigurationPageActionGroup" stepKey="goToAdobeStockConfigurationPage"/>
```
Comment (rogyar) [inline AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: Please, check my comment here https://github.com/magento/adobe-stock-integration/pull/239/files#r305875535
If it's possible, we should set the configuration values in a preferred way in order to avoid additional overhead.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="UserSearchesStockImagesByKeywords">
```
Comment (rogyar) [inline AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:11]: The test contains many actions that may be reused within the scope of other tests. However, until they are hardcoded in the test, they cannot be reusable. The best practice is to make the test consist of action groups only. At the same time, the action groups should be granular (small) to match the "reusable" principle. 
The main idea is another test can pick up a set of existing action groups and use them in the test body. 
Please, check my comments for this test https://github.com/magento/adobe-stock-integration/pull/239/files#diff-f63ad1182db6cace7d59611da116115fR40 and recent MFTF implementations https://github.com/magento/magento-functional-tests-migration/pull/707/files for examples

**Review by rogyar — CHANGES_REQUESTED** [review]

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml (hunk 6 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml]: @rogyar I've checked this pull request yesterday, it different action group with different behavior.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 8 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:11]: @rogyar All action moved to action group, in test I wrote only assertions.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 7 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: @rogyar If you record via the config:set - adobe api key and private key will be hardcoded, it is not suitable for our case.

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml (hunk 5 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml]: @rogyar For example in action group AdminCatalogSearchTermActionGroup we have several action groups. In all project we have many action groups which has several action group per file.

Comment (rogyar) [reply AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AdminOpenAdobeStockSearchModalActionGroup.xml]: Yes, that's true. Unfortunately, the tests that have been created in the very beginning are not the best example of best practices. From that moment we are trying to improve the best practices considering mistakes made in the past. It's like underscores for private/protected variables in the initial version of the Magento 2 core codebase and other similar examples. You can check the following wiki for short description of the best practices https://github.com/magento/magento-functional-tests-migration/wiki/Best-practices

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 7 above)
Comment (rogyar) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: Makes sense. However, we can have as many key pairs using different entities as we need. So, in this case no matter where exactly the values will be hardcoded in a test body or in the entities set. 

If we still need to put dynamic values in every test we can try passing `apiKey` and `privateKey` to a separate action group and calling `config:set` from the action group.

Comment (rogyar) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml]: Makes sense. However, we can have as many key pairs using different entities as we need. So, in this case no matter where exactly the values will be hardcoded in a test body or in the entities set.

If we still need to put dynamic values in every test we can try passing apiKey and privateKey to a separate action group and calling config:set from the action group.

**Comment by engcom-Golf** [comment]
HI @rogyar hi @hws47a Can you please review new changes ? , seems like now everything looks good.

**Comment by rogyar** [comment]
Hi @engcom-Golf. Awesome, thank you!

**Review by rogyar — APPROVED** [review]

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            </actionGroup>
+        </after>
+        <actionGroup ref="AdminOpenAdobeStockSearchModalActionGroup" stepKey="openSearchModal"/>
+        <grabAttributeFrom selector="{{AdobeStockSection.firstImageAfterSearch}}" userInput="src"
+                           stepKey="getUrlFromFirstImageWithoutSearch"/>
+        <actionGroup ref="AdminSearchImagesOnModalActionGroup" stepKey="searchForPopular">
+            <argument name="query" value="popular"/>
+        </actionGroup>
+        <!-- Assert that User searches for "popular" using keyboard input and mouse navigation and sees "that grid was updated by new images"-->
+        <dontSeeElement selector="{{AdobeStockSection.imageSrc($getUrlFromFirstImageWithoutSearch)}}"
```
Comment (hws47a) [inline AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:41]: what about a check that any item (or all 32 items) does exist on the page? Here we only check that the previous image is not here.

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AssertImageNotFoundAfterSearchFirstSymbolsActionGroup.xml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AssertImageNotFoundAfterSearchFirstSymbols">
+        <arguments>
+            <argument name="src" type="string" defaultValue=""/>
+        </arguments>
+        <dontSeeElement selector="//div[@class='masonry-image-column' and contains(@data-repeat-index, '0')]//img[@src='{{src}}']" stepKey="checkFirstPopularImageNotFoundAfterSearchFirstSymbols"/>
+        <dontSeeElement selector="//aside[contains(@class, 'modal-popup')]//div[@class='modal-content']//div[text()='Something went wrong.']" stepKey="errorModalIsAbsentAfterSearchFirstSymbols"/>
+        <dontSeeElement selector="//div[contains(@id, 'adobe-stock-images-search-modal')]//div[@class='spinner']" stepKey="spinnerIsAbsentAfterSearchFirstSymbols"/>
```
Comment (hws47a) [inline AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AssertImageNotFoundAfterSearchFirstSymbolsActionGroup.xml]: all of the selectors should be in a section file

**Review by hws47a — DISMISSED** [review]

File: AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AssertImageNotFoundAfterSearchFirstSymbolsActionGroup.xml (hunk 10 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/ActionGroup/AssertImageNotFoundAfterSearchFirstSymbolsActionGroup.xml]: ok, thanks for some reason i forget for this.

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 9 above)
Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:41]: @hws47a good idea, maybe i add action group that checks all grid if it contains images ?

**Comment by engcom-Golf** [comment]
Hi @hws47a thank you for you reviews comment, could you please review new changes ?

File: AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml (hunk 9 above)
Comment (hws47a) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:41]: There is one in #239, it's expected to be merged to develop branch soon.

Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:41]: @hws47a did you mean this group ? **AdminAssertAdobeStockThumbnailsNumberActionGroup** 
however i implemented new one that chekcs total count of images, can we use  this action group or we should use **AdminAssertAdobeStockThumbnailsNumberActionGroup** ?

Comment (engcom-Golf) [reply AdobeStockImageAdminUi/Test/Mftf/Test/UserSearchesStockImagesByKeywords.xml:41]: oh.. now i can see) your solution much better)

**Comment by engcom-Golf** [comment]
HI @hws47a I refactored to using your action groups, but i have one issue with action groups to open AdobeStockModal, this code ->  
```
<click selector="{{MediaGallerySection.Browse}}" stepKey="clickBrowse" />
        <waitForPageLoad stepKey="waitForGalleryOpen" />
```
Codeception not found any browse button and i'm too, also i change the selector for insert image because test fails and don't see icon element is this good for you ?

**UPDATED:** 
Are you disable WYSIWYG ?
https://devdocs.magento.com/mftf/docs/getting-started.html
In the WYSIWYG Options section set the Enable WYSIWYG Editor option to Disabled Completely.

**Review by sidolov — APPROVED** [review]
